flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java

```diff
@@ -0,0 +1,609 @@
+			removeContainerRequest(pendingRequest);
+
+			numAccepted++;
+		}
+		numPending -= numAccepted;
+
+		int numExcess = 0;
+		while (containerIterator.hasNext()) {
+			returnExcessContainer(containerIterator.next());
+			numExcess++;
+		}
+
+		log.info("Accepted {} requested containers, returned {} excess containers, {} pending container requests of resource {}.",
+			numAccepted, numExcess, numPending, resource);
+	}
+
+	private int getNumRequestedNotAllocatedWorkers() {
+		return requestResourceFutures.values().stream().mapToInt(Queue::size).sum();
+	}
+
+	private int getNumRequestedNotAllocatedWorkersFor(TaskExecutorProcessSpec taskExecutorProcessSpec) {
+		return requestResourceFutures.getOrDefault(taskExecutorProcessSpec, new LinkedList<>()).size();
+	}
+
+	private void removeContainerRequest(AMRMClient.ContainerRequest pendingContainerRequest) {
+		log.info("Removing container request {}.", pendingContainerRequest);
+		resourceManagerClient.removeContainerRequest(pendingContainerRequest);
+	}
+
+	private void returnExcessContainer(Container excessContainer) {
+		log.info("Returning excess container {}.", excessContainer.getId());
+		resourceManagerClient.releaseAssignedContainer(excessContainer.getId());
+	}
+
+	private void startTaskExecutorInContainer(Container container, TaskExecutorProcessSpec taskExecutorProcessSpec, ResourceID resourceId, CompletableFuture<YarnWorkerNode> requestResourceFuture) {
+		final YarnWorkerNode yarnWorkerNode = new YarnWorkerNode(container, resourceId);
+
+		try {
+			// Context information used to start a TaskExecutor Java process
+			ContainerLaunchContext taskExecutorLaunchContext = createTaskExecutorLaunchContext(
+				resourceId,
+				container.getNodeId().getHost(),
+				taskExecutorProcessSpec);
+
+			nodeManagerClient.startContainerAsync(container, taskExecutorLaunchContext);
+			requestResourceFuture.complete(yarnWorkerNode);
+		} catch (Throwable t) {
+			requestResourceFuture.completeExceptionally(t);
+		}
+	}
+
+	private Collection<AMRMClient.ContainerRequest> getPendingRequestsAndCheckConsistency(Resource resource, int expectedNum) {
+		final Collection<Resource> equivalentResources = taskExecutorProcessSpecContainerResourceAdapter.getEquivalentContainerResource(resource, matchingStrategy);
+		final List<? extends Collection<AMRMClient.ContainerRequest>> matchingRequests =
+			equivalentResources.stream()
+				.flatMap(equivalentResource -> resourceManagerClient.getMatchingRequests(
+					RM_REQUEST_PRIORITY,
+					ResourceRequest.ANY,
+					equivalentResource).stream())
+				.collect(Collectors.toList());
+
+		final Collection<AMRMClient.ContainerRequest> matchingContainerRequests;
+
+		if (matchingRequests.isEmpty()) {
+			matchingContainerRequests = Collections.emptyList();
+		} else {
+			final Collection<AMRMClient.ContainerRequest> collection = matchingRequests.get(0);
+			matchingContainerRequests = new ArrayList<>(collection);
+		}
```

Comment:
I think it would be a valid issue. In the [description](https://github.com/apache/hadoop/blob/deb35a32bafdd3065e3c2f243d84ef79209838e9/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/main/java/org/apache/hadoop/yarn/client/api/AMRMClient.java#L684) of this interface, it seems it will return all the requests which could fit in the given resource. However, in its [implementation](https://github.com/apache/hadoop/blob/deb35a32bafdd3065e3c2f243d84ef79209838e9/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/main/java/org/apache/hadoop/yarn/client/api/impl/AMRMClientImpl.java#L734), it will first try to return the exactly matching requests. If not found, it will regress to the aforementioned logic.

Recently in [FLINK-19324](https://issues.apache.org/jira/browse/FLINK-19324), we found that we perhaps need to use priority to map the different requests. If so, this issue should be also fixed from my understanding. Anyway, I tend to put this issue out of the scope of this PR. WDYT?

Reply:
Let's then create a follow up issue to keep track of it.